rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+This should be driven by real world use cases. For example:
+
+1. Ignoring existing checkpoints
+1. Start at the beginning or end of existing files, optionally taking into
+   account factors like mtime
+1. Start at the beginning or end of files added while we're watching (this can
+   be tricky)
+1. Ordering which of the above concerns take precedence
+
+I would suggest a config like the following:
+
+* `ignore_checkpoints = true|false` (still write but don't read existing checkpoints)
+* `read_from = beginning|end` (where to start if there's no checkpoint or
+    they're being ignored)
+* `skip_older_than = duration` (relevant when `read_from = beginning`, seek to
+    end based on mtime)
+* Always start at the beginning of files added while we're watching (it's hard
+    to tell a `mv` from a create and write, so don't rely on seeing an empty file
+    first to get all the data)
+
+If we adopt this, we can also implement a solution where we don't hold open file
+handles for files that match `skip_older_than`, since this has caused some
+issues. Naively, we need the open handle to attempt reads in case new data is
+written to the file, but we could also implement that as a new state of
+`FileWatcher` that stashes the inital size and periodically checks the file
+metadata (mtime and size) to see if it should start reading at what was the end.
+
+### General tweaks
+
+In addition to extracting these various subcomponents, there are some other
+relatively simple changes we can make to help address the issues we discussed.
+
+#### File discovery and checkpointing
+
+We currently use the outdated `glob_minimum_cooldown` config option to determine
+how often to do both of these tasks. We should switch them to their own
+independent config options and allow them to be disabled (e.g. a batch use case
+does not need to continuously look for new files).
+
+We should also move them into their own periodic tasks, outside of the main read
+loop. This should give us better performance and help avoid bad behavior in
+situations where either is expensive (e.g. discovering millions of files on
+EBS).
+
+#### Read concurrency
+
+With most of the other concerns pulled out, we can afford to adjust the main
+read loop to allow for some new capabilities. The most interesting is concurrent
+reads, but it will require a bit of experimentation before we're able to
+determine if it's worthwhile. There are a few possible approaches:
+
+1. Dispatch reads to an explicit threadpool
+1. Spawn a limited number of blocking tokio tasks
+1. Implement something with `iouring`
+
+The first two both introduce the questions of sizing and the ability of the
+underlying file system to enable concurrent access in a way that actually adds
+performance. We would need to test a wide variety of scenarios to evaluate the
+best path for either.
+
+Much more interesting, but also limited, would be to build on top of `iouring`.
```

Comment:
I don't really think this is a practical option at this time.

Reply:
I know you don't have a crystal ball, but what kind of timeline are we talking about before this becomes a practical advantage?

Reply:
It would be an advantage right now, but it's (1) not available everywhere, and (2) a lot of work to implement that will likely need redone as library interfaces to it mature past their current experimental stage.